include/fiction/algorithms/physical_design/graph_oriented_layout_design.hpp

```diff
@@ -0,0 +1,1523 @@
+                    }
+                }
+            }
+
+            timeout_limit_reached =
+                std::none_of(ssg_vec.cbegin(), ssg_vec.cend(), [](const auto& ssg) { return ssg.frontier_flag; });
+
+            const auto end = std::chrono::high_resolution_clock::now();
+            const auto duration_ms =
+                static_cast<uint64_t>(std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count());
+
+            if (duration_ms >= timeout)
+            {
+                timeout_limit_reached = true;
+            }
+        }
+
+        return best_lyt;
+    }
+
+  private:
+    /**
+     * Alias for an obstruction layout based on the given layout type.
+     */
+    using ObstrLyt = obstruction_layout<Lyt>;
+    /**
+     * The network to be placed and routed.
+     */
+    tec_nt ntk;
+    /**
+     * Parameters.
+     */
+    graph_oriented_layout_design_params ps;
+    /**
+     * Statistics.
+     */
+    graph_oriented_layout_design_stats& pst;
+    /**
+     * Timeout limit (in ms).
+     */
+    uint64_t timeout;
+    /**
+     * Start time.
+     */
+    std::chrono::time_point<std::chrono::high_resolution_clock> start;
+    /**
+     * Number of search space graphs.
+     */
+    uint64_t num_search_space_graphs;
+    /**
+     * Vector of search space graphs.
+     */
+    std::vector<search_space_graph<ObstrLyt>> ssg_vec;
+    /**
+     * Count evaluated paths in the search space graphs.
+     */
+    uint64_t num_evaluated_paths{0ul};
+    /**
+     * Keep track of the maximum number of placed nodes.
+     */
+    uint64_t max_placed_nodes{0ul};
+    /**
+     * Current best solution w.r.t. area.
+     */
+    uint64_t best_solution{std::numeric_limits<uint64_t>::max()};
+    /**
+     * Current best solution w.r.t. area after relocating POs.
+     */
+    uint64_t best_optimized_solution{100000};
+    /**
+     * Flag indicating if an initial solution was already found, and that other search space graphs should be pruned.
+     */
+    bool improv_mode{false};
+    /**
+     * Checks if there is a path between the source and destination tiles in the given layout.
+     * Optionally returns the path if specified.
+     *
+     * @param layout The layout to be checked.
+     * @param src The source tile.
+     * @param dest The destination tile.
+     * @param src_is_new_pos Flag indicating if the source is a new position and should be checked if empty. Defaults to
+     * `false`.
+     * @param dest_is_new_pos Flag indicating if the destination is a new position and should be checked if empty.
+     * Defaults to `false`.
+     * @return A path from `src` to `dest` if one exist, `std::nullopt` otherwise.
+     */
+    [[nodiscard]] std::optional<layout_coordinate_path<ObstrLyt>>
+    check_path(const ObstrLyt& layout, const tile<ObstrLyt>& src, const tile<ObstrLyt>& dest,
+               const bool src_is_new_pos = false, const bool dest_is_new_pos = false) noexcept
```

Comment:
There is an ongoing debate in the C++ community about whether to ban `bool` arguments because they make the code so unreadable. On the caller's end, it is simply unclear what `check_path(... , true, false)` means. Plus, it is easy to accidentally swap the parameters. Strongly typed `enum`s are preferrable, i.e. `enum class` for each parameter type. Here, you might even merge them into one, e.g.
```c++
enum class new_po_type
{
NONE,
PI,
PO,
PI_AND_PO
};
```
where these names are to be debated, should just convey the message.